### ISSUE TITLE
Logging and String Interpolation Errors in gradient

### ISSUE THREAD

**Issue by jessebett** [FluxML/Zygote.jl#560]
Minimum Error:

```julia
using Zygote
using Logging

f(x) = 2x

# works
Zygote.gradient(1.) do x
  loss = f(x)
  return loss
end

# works with vanilla printing
Zygote.gradient(1.) do x
  loss = f(x)
  println(loss)
  print(loss)
  return loss
end

# fails with string interpolation
# ERROR: Can't differentiate foreigncall expression
# also takes a long time
Zygote.gradient(1.) do x
  loss = f(x)
  print("loss: $loss")
  return loss
end

# fails with logging
# ERROR: Compiling Tuple{var"#43#44",Float64}: try/catch is not supported.
Zygote.gradient(1.) do x
  loss = f(x)
  @info loss
  return loss
end
```

Relevant Stacktraces:
```
ERROR: Can't differentiate foreigncall expression
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] String at ./strings/string.jl:39 [inlined]
 [3] (::typeof(∂(String)))(::Nothing) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/interface2.jl:0
 [4] print_to_string at ./strings/io.jl:140 [inlined]
 [5] (::typeof(∂(print_to_string)))(::Nothing) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/interface2.jl:0
 [6] (::Zygote.var"#173#174"{typeof(∂(print_to_string)),Tuple{Tuple{Nothing,Nothing}}})(::Nothing) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/lib/lib.jl:159
 [7] #334#back at /Users/jesse/.julia/packages/ZygoteRules/6nssF/src/adjoint.jl:49 [inlined]
 [8] string at ./strings/io.jl:177 [inlined]
 [9] #41 at ./REPL[29]:3 [inlined]
 [10] (::typeof(∂(#41)))(::Float64) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/interface2.jl:0
 [11] (::Zygote.var"#38#39"{typeof(∂(#41))})(::Float64) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/interface.jl:36
 [12] gradient(::Function, ::Float64) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/interface.jl:45
 [13] top-level scope at REPL[29]:1

ERROR: Compiling Tuple{var"#43#44",Float64}: try/catch is not supported.
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] instrument(::IRTools.Inner.IR) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/reverse.jl:89
 [3] #Primal#16 at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/reverse.jl:170 [inlined]
 [4] Type at ./none:0 [inlined]
 [5] #Adjoint#32(::Nothing, ::Bool, ::Type{Zygote.Adjoint}, ::IRTools.Inner.IR) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/reverse.jl:283
 [6] (::Core.var"#kw#Type")(::NamedTuple{(:varargs, :normalise),Tuple{Nothing,Bool}}, ::Type{Zygote.Adjoint}, ::IRTools.Inner.IR) at ./none:0
 [7] _lookup_grad(::Type) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/emit.jl:101
 [8] #s3631#1830 at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/interface2.jl:21 [inlined]
 [9] #s3631#1830(::Any, ::Any, ::Any) at ./none:0
 [10] (::Core.GeneratedFunctionStub)(::Any, ::Vararg{Any,N} where N) at ./boot.jl:524
 [11] (::Zygote.var"#38#39"{typeof(∂(#43))})(::Float64) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/interface.jl:36
 [12] gradient(::Function, ::Float64) at /Users/jesse/.julia/packages/Zygote/KNUTW/src/compiler/interface.jl:45
 [13] top-level scope at REPL[32]:1
```

**Comment by jessebett** [comment]
@mcabbott pointed out this is a duplicate of https://github.com/FluxML/Zygote.jl/issues/386 for the logging compatibility.

**Comment by mcabbott** [comment]
Some things are easily fixed by `@nograd`, like these two:
```julia
Zygote.@nograd string, repr

Zygote.gradient(1.) do x
  loss = f(x)
  print("loss: $loss") # calls string
  return loss
end

Zygote.gradient(1.) do x
  loss = f(x)
  @show loss  # calls repr
  return loss
end
```

**Comment by mcabbott** [comment]
xref #548 about println, removed in #555 because of #554.

**Comment by jessebett** [comment]
@willtebbutt was there a reason you removed the `@nograd` for prints in #554 ?

**Comment by willtebbutt** [comment]
Yeah, because they're defined [here](https://github.com/FluxML/Zygote.jl/blob/17ca911b82134c4a765822cd2b7ee19e959cc8e4/src/lib/base.jl#L1). I accidentally defined them for a second time in a previous PR when I was adding the Core.print stuff.

**Comment by jessebett** [comment]
Okay just making sure we're not going in circles. Thanks
